plugins/chanbackup.c

```diff
@@ -0,0 +1,393 @@
+	}
+	/*~ We actually need to sync the *directory itself* to make sure the
+	 * file exists!  You're only allowed to open directories read-only in
+	 * modern Unix though. */
+	fd = open(".", O_RDONLY);
+	if (fd < 0) {
+		plugin_log(p, LOG_DBG, "Opening: %s", strerror(errno));
+	}
+	if (fsync(fd) != 0) {
+		unlink_noerr("scb.tmp");
+	}
+	close(fd);
+	unlink_noerr("emergency.recover");
+	rename("scb.tmp", "emergency.recover");
+}
+
+static struct command_result *after_staticbackup(struct command *cmd,
+					         const char *buf,
+					         const jsmntok_t *params,
+					         void *cb_arg UNUSED)
+{
+	const jsmntok_t *scbs = json_get_member(buf, params, "scb");
+	plugin_log(cmd->plugin, LOG_INFORM, "Updating the SCB");
+
+	update_scb(cmd->plugin, json_strdup(tmpctx, buf, scbs));
+	return notification_handled(cmd);
+}
+
+static struct command_result *json_state_changed(struct command *cmd,
+					         const char *buf,
+					         const jsmntok_t *params)
+{
+	const jsmntok_t *notiftok = json_get_member(buf,
+                                                    params,
+                                                    "channel_state_changed"),
+	*statetok = json_get_member(buf, notiftok, "new_state");
+
+	/* FIXME: I wanted to update the file on CHANNELD_AWAITING_LOCKIN,
+	 * But I don't get update for it, maybe because there is no previous_state,
+	 * also apparently `channel_opened` gets published when *peer* funded a channel with us?
+	 * So, is their no way to get a notif on CHANNELD_AWAITING_LOCKIN? */
+	if (json_tok_streq(buf, statetok, "CLOSED") ||
+		json_tok_streq(buf, statetok, "CHANNELD_NORMAL")) {
+
+		struct out_req *req;
+		req = jsonrpc_request_start(cmd->plugin,
+                                            cmd,
+                                            "staticbackup",
+                                            after_staticbackup,
+                                            &forward_error,
+                                            NULL);
+
+		return send_outreq(cmd->plugin, req);
+	}
+
+	return notification_handled(cmd);
+}
+
+static const char *init(struct plugin *p,
+			const char *buf UNUSED,
+			const jsmntok_t *config UNUSED)
+{
+	const char *scb_buf;
+	rpc_scan(p, "staticbackup",
+		 take(json_out_obj(NULL, NULL, NULL)),
+		 "{scb:%}", JSON_SCAN(json_to_str, &scb_buf));
+
+	rpc_scan(p, "makesecret",
+		 take(json_out_obj(NULL, "info", "scb secret")),
+		 "{secret:%}", JSON_SCAN(json_to_secret, &secret));
+
+	plugin_log(p, LOG_DBG, "Chanbackup Initialised!");
+
+	/* flush the tmp file, if exists */
+	unlink_noerr("scb.tmp");
+
+	maybe_create_new_scb(p, scb_buf);
+
+	return NULL;
+}
+
+static const struct plugin_notification notifs[] = {
+	{
+		"channel_state_changed",
+		json_state_changed,
+	}
+};
+
+static const struct plugin_command commands[] = { {
+		"emergencyrecover",
+		"recovery",
+		"Populates the DB with stub channels",
+		"returns stub channel-id's on completion",
+		recover,
```

Comment:
I always like the function for a command to be json_COMMANDNAME, in this case json_emergencyrecover.  That way I can easily find it!